api/handlers/videostream.go

```diff
@@ -35,54 +35,340 @@ LICENSE
+	"github.com/ausocean/openfish/datastore"
 
 	"github.com/gofiber/fiber/v2"
 )
 
 // VideoStreamResult  describes the JSON format for video streams in API responses.
 // Fields use pointers because they are optional (this is what the format URL param is for).
 type VideoStreamResult struct {
-	ID        *int           `json:"id,omitempty"`
-	Datetime  *time.Time     `json:"datetime,omitempty"`
-	Duration  *time.Duration `json:"duration,omitempty"`
-	StreamUrl *string        `json:"streamUrl,omitempty"`
+	ID            *int       `json:"id,omitempty"`
+	StartTime     *time.Time `json:"startTime,omitempty"`
+	EndTime       *time.Time `json:"endTime,omitempty"`
+	StreamUrl     *string    `json:"stream_url,omitempty"`
+	CaptureSource *int64     `json:"capturesource,omitempty"`
+}
+
+// FromVideoStream creates a VideoStreamResult from a model.VideoStream and key, formatting it according to the requested format.
+func FromVideoStream(videoStream *model.VideoStream, id int, format *api.Format) VideoStreamResult {
+	var result VideoStreamResult
+	if format.Requires("id") {
+		result.ID = &id
+	}
+	if format.Requires("start_time") {
+		result.StartTime = &videoStream.StartTime
+	}
+	if format.Requires("end_time") {
+		result.EndTime = videoStream.EndTime
+	}
+	if format.Requires("stream_url") {
+		result.StreamUrl = &videoStream.StreamUrl
+	}
+	if format.Requires("capturesource") {
+		result.CaptureSource = &videoStream.CaptureSource
+	}
+	return result
 }
 
 // GetVideoStreamsQuery describes the URL query parameters required for the GetVideoStreams endpoint.
 type GetVideoStreamsQuery struct {
-	TimeSpan      *string `query:"timespan"`       // Optional. TODO: choose more appropriate type.
-	CaptureSource *int64  `query:"capture_source"` // Optional.
+	CaptureSource *int64 `query:"capturesource"` // Optional.
+	TimeSpan      *struct {
+		Start time.Time `query:"start"`
+		End   time.Time `query:"end"`
+	} `query:"timespan"` // Optional.
 	api.LimitAndOffset
-	api.Format
+}
+
+// CreateVideoStreamBody describes the JSON format required for the CreateVideoStream endpoint.
+//
+// ID is omitted because it is chosen automatically.
+type CreateVideoStreamBody struct {
+	StartTime     time.Time `json:"startTime"`
+	EndTime       time.Time `json:"endTime"`
+	StreamUrl     string    `json:"stream_url"`
+	CaptureSource int64     `json:"capturesource"`
+}
+
+// StartVideoStreamBody describes the JSON format required for the StartVideoStream endpoint.
+//
+// ID is omitted because it is chosen automatically.
+// Datetime is omitted because it uses the current time.
+// Duration is omitted because it will be set once the stream concludes.
+type StartVideoStreamBody struct {
+	StreamUrl     string `json:"stream_url"`
+	CaptureSource int64  `json:"capturesource"`
+}
+
+// UpdateVideoStreamBody describes the JSON format required for the UpdateVideoStream endpoint.
+type UpdateVideoStreamBody struct {
+	StartTime     *time.Time `json:"startTime"`     // Optional.
```

Comment:
TimeSpan ?

Reply:
StartTime and EndTime are optional - in a single request we might only update the start time, or only update the end time. If we use a *TimeSpan, it has to be either nil, or both have to be supplied. So not appropriate in my opinion.